File: frontend/src/types/absence.ts
import { MatiereResponse } from "./matiere";
import { UtilisateurResponse } from "./user";

export type StatutAbsence = "en_attente" | "valide" | "rejete";

export interface Absence {
  id: number;
  enseignant_id: number;
  matiere_id: number;
  date_absence: string;
  motif: string;
  justificatif: string | null;
  statut: StatutAbsence;
  created_at: string;
  updated_at: string;
}

export interface AbsenceResponse extends Absence {
  enseignant?: {
    id: number;
    nom: string;
    prenom: string;
    email: string;
  } & Partial<UtilisateurResponse>;
  matiere?: MatiereResponse;
}

export interface CreateAbsencePayload {
  matiere_id: number;
  date_absence: string;
  motif: string;
  justificatif?: File;
}

export interface UpdateAbsencePayload {
  matiere_id?: number;
  date_absence?: string;
  motif?: string;
  justificatif?: File;
}

File: backend/app/schemas/absence.py
from typing import Optional
from datetime import date, datetime
from pydantic import BaseModel, Field, ConfigDict, field_validator
from app.models.enums import StatutAbsence
from app.schemas.matiere import MatiereSimple
from app.schemas.utilisateur import UtilisateurSimple

class AbsenceBase(BaseModel):
    matiere_id: int = Field(..., description="Identifiant de la matière")
    date_absence: date = Field(..., description="Date de l'absence")
    motif: str = Field(..., min_length=1, description="Motif de l'absence")
    
    @field_validator('motif')
    @classmethod
    def validate_motif(cls, v):
        if not v or len(v.strip()) == 0:
            raise ValueError('Le motif de l\'absence ne peut pas être vide')
        return v

class AbsenceCreate(AbsenceBase):
    pass

class AbsenceUpdate(BaseModel):
    matiere_id: Optional[int] = None
    date_absence: Optional[date] = None
    motif: Optional[str] = None

class AbsenceSimple(AbsenceBase):
    id: int
    enseignant_id: int
    enseignant: Optional[UtilisateurSimple] = None
    matiere: Optional[MatiereSimple] = None
    justificatif: Optional[str] = None
    statut: StatutAbsence
    model_config = ConfigDict(from_attributes=True)

class AbsenceResponse(AbsenceBase):
    id: int
    enseignant_id: int
    enseignant: Optional[UtilisateurSimple] = None
    matiere: Optional[MatiereSimple] = None
    justificatif: Optional[str] = None
    statut: StatutAbsence
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)
